### ISSUE TITLE
Parse error on very simple json document

### ISSUE THREAD

**Issue by Gei0r** [biojppm/rapidyaml#52]
As I understand it, [yaml is a superset of json](https://yaml.org/spec/1.2/spec.html#id2759572), but rapidyml throws an error when trying to parse this very simple document

```C++
#include <ryml.hpp>
int main() {
    ryml::Tree t = ryml::parse("{\"a\":\"b\"}");
}
```
Output:
```
ERROR parsing yml: parse error
line 1: '{"a":"b"}' (sz=9)
             ^~~~~  (cols 5-10)
```

**Comment by biojppm** [comment]
It is not valid YAML; the colon is missing a space after it, and the correct format is `{"a": "b"}`. 

Although all the online validators say this is valid, they are wrong. [See this](https://stackoverflow.com/questions/42124227/why-does-the-yaml-spec-mandate-a-space-after-the-colon). 

So make sure to emit `"key": "value"` in your emitter, instead of `"key":"value"`. 

You could argue that ryml could make an effort to try to read this, but my position here is to avoid the complexity, to prevent the code from becoming fragile and slow. It's a tradeoff. Maybe in the future this can be implemented as ryml matures but for now it's off limits.

**Comment by Gei0r** [comment]
That's sad, because I now need a json parser in addition to a yaml parse. I hoped I could get away with just using rapidyaml.

But you're right, it's a problem with the yaml spec, not rapidyaml.

**Comment by biojppm** [comment]
To be fair, it's not the first time this appears; I've had other people complaining about the same exact problem. So maybe it would be worth it to try to address it somehow. I do think that doing this in the parser would make it even more complex thant it is, so I really am 

But I'm thinking that a json->yaml preprocessing step before parsing would address this case. It would just add spaces after the `:` and maybe other required cleanup. Of course, this would need to be explicitly done. Something like:
```c++
char buf[100];
csubstr json = "{\"a\":\"b\"}";
substr yaml = buf;

ryml::preprocess_json(json, &yaml);  // <----- thinking of adding this function
// now you have a buffer with valid yaml - note the space: 
std::cout << yaml << "\n"; // {"a": "b"}

// ... which you can parse:
ryml::Tree t = ryml::parse(yaml);
std::cout << t["a"] << "\n"; // b
```

**Comment by biojppm** [comment]
What do you think?

**Comment by Gei0r** [comment]
Yeah, such a function would be very helpful.

**Comment by biojppm** [comment]
@Gei0r I just merged the JSON preprocessor to master . Have a try, and feel free to reopen this if you see something off.

**Comment by Gei0r** [comment]
@biojppm I've only now gotten around to it.

I think the include guard in `preprocess.hpp` is messed up:
```c++
#ifndef _C4_YML_EMIT_HPP_
#define _C4_YML_EMIT_HPP_

// end of file:
#endif /* _C4_YML_COMMON_HPP_ */
```

**Comment by biojppm** [comment]
Indeed, thanks for the catch.

**Comment by Gei0r** [comment]
The implementation can't correctly handle the following inputs:

```json
{"a":"b","c":42,"d":"e"}     # doesn't put a space after "d":
{"a":"b","c":{},"d":"e"}     # same as before
{"a":"b","c":false,"d":"e"}  # doesn't put a space after "c":
```

Full test case:

```c++
#include <string>
#include <ryml_std.hpp>
#include <c4/yml/preprocess.hpp>
#include <ryml.hpp>

#include <iostream>

int main()
{
    std::string in = R"({"a":"b","c":{},"d":"e"})";
    std::string yaml = ryml::preprocess_json<std::string>(c4::to_csubstr(in));

    std::cout << "BEFORE: " << in << "\n";
    std::cout << "AFTER:  " << yaml << "\n\n";
    ryml::Tree t = ryml::parse(c4::to_substr(yaml));
    std::cout << "done\n";
}
```
Output:
```
BEFORE: {"a":"b","c":{},"d":"e"}
AFTER:  {"a": "b","c": {},"d":"e"}

1:22: (21B):ERROR: ERROR parsing yml: parse error
line 1: '{"a": "b","c": {},"d":"e"}' (sz=26)
                              ^~~~~  (cols 22-27)
```

**Comment by Gei0r** [comment]
By the way, the new error callback parameter "Location" is great!

**Comment by biojppm** [comment]
Thanks for reporting the problem; let me have a look at it.

**Comment by biojppm** [comment]
I appreciate that the repro is so detailed -- it does help me save some time. I'll try to investigate tomorrow or hopefully this evening.

**Comment by biojppm** [comment]
@Gei0r Can you have another go on this branch? Thanks for the patience, and please let me know if you find any other problems.

**Comment by biojppm** [comment]
I'm assuming this is fixed. Feel free to reopen if you find more issues.

**Comment by Gei0r** [comment]
Hi, thank you, it works now.

However, there's an annoying dependency on include order. If I include files in the following order:
```c++
#include <ryml.hpp>
#include <ryml_std.hpp>
#include <c4/yml/preprocess.hpp>
```
I get the following error:
```
[ 50%] Building CXX object CMakeFiles/ryml_ausprobieren.dir/main2.cpp.obj

In file included from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/yml.hpp:8,
                 from D:/User/Adrian/Programmierung/ryml/install/include/ryml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:1:
D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/preprocess.hpp: In instantiation of 'c4::substr c4::yml::detail::preprocess_into_container(c4::csubstr, CharContainer*) [with size_t (* PP)(c4::csubstr, c4::substr) = c4::yml::preprocess_json; CharContainer = std::__cxx11::basic_string<char>; c4::substr = c4::basic_substring<char>; c4::csubstr = c4::basic_substring<const char>]':
D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/preprocess.hpp:77:62:   required from 'c4::substr c4::yml::preprocess_json(c4::csubstr, CharContainer*) [with CharContainer = std::__cxx11::basic_string<char>; c4::substr = c4::basic_substring<char>; c4::csubstr = c4::basic_substring<const char>]'
D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/preprocess.hpp:86:20:   required from 'CharContainer c4::yml::preprocess_json(c4::csubstr) [with CharContainer = std::__cxx11::basic_string<char>; c4::csubstr = c4::basic_substring<const char>]'
D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:13:77:   required from here
D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/preprocess.hpp:33:36: error: no matching function for call to 'to_substr(std::__cxx11::basic_string<char>&)'
   33 |     size_t sz = PP(input, to_substr(*out));
      |                           ~~~~~~~~~^~~~~~
In file included from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/common.hpp:5,
                 from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/tree.hpp:6,
                 from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/yml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/install/include/ryml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:1:
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1464:15: note: candidate: 'c4::substr c4::to_substr(char*)'
 1464 | inline substr to_substr(char *s)
      |               ^~~~~~~~~
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1464:31: note:   no known conversion for argument 1 from 'std::__cxx11::basic_string<char>' to 'char*'
 1464 | inline substr to_substr(char *s)
      |                         ~~~~~~^
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1513:15: note: candidate: 'c4::substr c4::to_substr(c4::substr)'
 1513 | inline substr to_substr(substr s)
      |               ^~~~~~~~~
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1513:32: note:   no known conversion for argument 1 from 'std::__cxx11::basic_string<char>' to 'c4::substr' {aka 'c4::basic_substring<char>'}
 1513 | inline substr to_substr(substr s)
      |                         ~~~~~~~^
In file included from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/yml.hpp:8,
                 from D:/User/Adrian/Programmierung/ryml/install/include/ryml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:1:
D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/preprocess.hpp:39:33: error: no matching function for call to 'to_substr(std::__cxx11::basic_string<char>&)'
   39 |         sz = PP(input, to_substr(*out));
      |                        ~~~~~~~~~^~~~~~
In file included from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/common.hpp:5,
                 from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/tree.hpp:6,
                 from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/yml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/install/include/ryml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:1:
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1464:15: note: candidate: 'c4::substr c4::to_substr(char*)'
 1464 | inline substr to_substr(char *s)
      |               ^~~~~~~~~
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1464:31: note:   no known conversion for argument 1 from 'std::__cxx11::basic_string<char>' to 'char*'
 1464 | inline substr to_substr(char *s)
      |                         ~~~~~~^
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1513:15: note: candidate: 'c4::substr c4::to_substr(c4::substr)'
 1513 | inline substr to_substr(substr s)
      |               ^~~~~~~~~
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1513:32: note:   no known conversion for argument 1 from 'std::__cxx11::basic_string<char>' to 'c4::substr' {aka 'c4::basic_substring<char>'}
 1513 | inline substr to_substr(substr s)
      |                         ~~~~~~~^
In file included from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/yml.hpp:8,
                 from D:/User/Adrian/Programmierung/ryml/install/include/ryml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:1:
D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/preprocess.hpp:41:21: error: no matching function for call to 'to_substr(std::__cxx11::basic_string<char>&)'
   41 |     return to_substr(*out).first(sz);
      |            ~~~~~~~~~^~~~~~
In file included from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/common.hpp:5,
                 from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/tree.hpp:6,
                 from D:/User/Adrian/Programmierung/ryml/install/include/c4/yml/yml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/install/include/ryml.hpp:4,
                 from D:/User/Adrian/Programmierung/ryml/ausprobieren/main2.cpp:1:
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1464:15: note: candidate: 'c4::substr c4::to_substr(char*)'
 1464 | inline substr to_substr(char *s)
      |               ^~~~~~~~~
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1464:31: note:   no known conversion for argument 1 from 'std::__cxx11::basic_string<char>' to 'char*'
 1464 | inline substr to_substr(char *s)
      |                         ~~~~~~^
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1513:15: note: candidate: 'c4::substr c4::to_substr(c4::substr)'
 1513 | inline substr to_substr(substr s)
      |               ^~~~~~~~~
D:/User/Adrian/Programmierung/ryml/install/include/c4/substr.hpp:1513:32: note:   no known conversion for argument 1 from 'std::__cxx11::basic_string<char>' to 'c4::substr' {aka 'c4::basic_substring<char>'}
 1513 | inline substr to_substr(substr s)
      |                         ~~~~~~~^
make[2]: *** [CMakeFiles/ryml_ausprobieren.dir/build.make:83: CMakeFiles/ryml_ausprobieren.dir/main2.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:96: CMakeFiles/ryml_ausprobieren.dir/all] Error 2
make: *** [Makefile:104: all] Error 2
```

It does work if `c4/yml/preprocess.hpp` is included first.

**Comment by Gei0r** [comment]
And another problem ;-)

```
BEFORE: {"a":"b","c":null,"d":"e"}
AFTER:  {"a": "b","c":null,"d": "e"}

1:14: (13B):ERROR: ERROR parsing yml: parse error
line 1: '{"a": "b","c":null,"d": "e"}' (sz=28)
                      ^~~~~~~~~~~~~~~  (cols 14-29)
```

**Comment by biojppm** [comment]
Ha. The second problem is an easy. I was lazy and assumed that the `true` and `false` tokens were the only valid literals, but of course there is still `null`.

**Comment by biojppm** [comment]
As for the header dependency problem...

Generally you have to include the std header before any other headers that use functions from it. This approach has served me well every time. For example, see [this example](https://github.com/biojppm/rapidyaml/blob/master/test/basic.cpp) or [this other example](https://github.com/biojppm/rapidyaml/blob/master/test/preprocess.cpp).

But I am confused by your report, which goes against what I just put above. Can you confirm?

**Comment by Gei0r** [comment]
I will create a new issue for the header dependency problem, so we don't get confused.

**Comment by Gei0r** [comment]
As for the preprocess functionality: It seems to work, but i stumbled across #65.

**Comment by biojppm** [comment]
Thanks for sticking through all the difficulties, and special thanks for reporting the problems. I'm closing this bug; again feel free to reopen for json-preprocessing issues.
